File: src/java/foundgine-providers/src/main/java/com/foundgine/providers/storage/sql/mutation/SqlMutationPlan.java
package com.foundgine.providers.storage.sql.mutation;

import com.foundgine.core.abstractions.FieldId;
import com.foundgine.core.execution.mutation.ProviderMutationPlan;
import com.foundgine.providers.storage.sql.query.SqlParameterBinding;
import java.util.*;

public final class SqlMutationPlan extends ProviderMutationPlan {
	private final String commandText;
	private final List<SqlParameterBinding> parameters;
	private final List<MutationReturnBinding> returnedFields;
	private final String fallbackCommandText;

	public SqlMutationPlan(String c, List<SqlParameterBinding> p, List<MutationReturnBinding> r, String f) {
		commandText = Objects.requireNonNull(c);
		parameters = List.copyOf(p);
		returnedFields = List.copyOf(r);
		fallbackCommandText = f;
	}

	public SqlMutationPlan(String c, List<SqlParameterBinding> p, List<MutationReturnBinding> r) {
		this(c, p, r, null);
	}

	public String commandText() {
		return commandText;
	}

	public List<SqlParameterBinding> parameters() {
		return parameters;
	}

	public List<MutationReturnBinding> returnedFields() {
		return returnedFields;
	}

	public String fallbackCommandText() {
		return fallbackCommandText;
	}
}

File: src/csharp/Foundgine.Providers/Storage/Sql/Mutation/SqlMutationPlan.cs
using Foundgine.Core.Execution.Mutation;
using Foundgine.Core.Semantic.Metadata;
using Foundgine.Core.Abstractions;
using Foundgine.Core.Semantic.Planning.Mutation;
using Foundgine.Providers.Storage.Sql.Query;

namespace Foundgine.Providers.Storage.Sql.Mutation;

public sealed record SqlMutationPlan(
    string CommandText,
    IReadOnlyList<SqlParameterBinding> Parameters,
    IReadOnlyList<MutationReturnBinding> ReturnedFields,
    string? FallbackCommandText = null) : ProviderMutationPlan;

public sealed record MutationReturnBinding(
    FieldId FieldId,
    string ResultName);

public sealed record SqlMutationBatchPlan(
    IReadOnlyList<SqlMutationPlan> Operations,
    IReadOnlyList<Foundgine.Core.Semantic.Planning.Mutation.MutationDependency> Dependencies)
    : Foundgine.Core.Execution.Mutation.ProviderMutationBatchPlan(Operations)
{
    public new IReadOnlyList<SqlMutationPlan> Operations { get; init; } = Operations;
}